perf(ivy): chain multiple property instructions

crisbeto opened the pull request:
Currently each property binding generates an instruction like this:

```
property('prop1', ctx.value1);
property('prop2', ctx.value2);
```

The problem is that we're repeating the call to `property` for each of the properties. Since the `property` instruction returns itself, we can chain all of the calls which is more compact and it looks like this:

```
property('prop1', ctx.value1)('prop2', ctx.value2);
```

These changes implement the chaining behavior for regular property bindings and for synthetic ones, however interpolated ones are still handled like before, because they use a separate instruction.

This PR resolves FW-1389.

crisbeto commented on packages/compiler/src/output/output_interpreter.ts:
```diff
@@ -195,6 +195,9 @@ class StatementInterpreter implements o.StatementVisitor, o.ExpressionVisitor {
     }
   }
+  visitInvokeFunctionChainExpr(stmt: o.InvokeFunctionChainExpr, ctx: _ExecutionContext): any {
+    throw new Error('Method not implemented.');
```
Note: from what I could tell this was only being used in ViewEngine so I haven't implemented it.

alfaproject commented on packages/compiler-cli/src/ngtsc/translator/src/translator.ts:
```diff
@@ -223,9 +223,25 @@ class ExpressionTranslatorVisitor implements ExpressionVisitor, StatementVisitor
         ast.args.map(arg => arg.visitExpression(this, context)));
-    if (ast.pure) {
-      ts.addSyntheticLeadingComment(expr, ts.SyntaxKind.MultiLineCommentTrivia, '@__PURE__', false);
+    ast.pure && this.setPureComment(expr);
```
You should use the more verbose version and you guys should enable that linting rule d:

alfaproject commented:
Would it be good to have this as some compiler configuration option?

I'm thinking that in dev mode, we might like to have each instruction separately for debugging purposes (breakpoints / stepping through), but not sure how common such need would be

devversion commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1383,6 +1418,11 @@ function instruction(
 
+function chainedInstruction(
+    span: ParseSourceSpan | null, reference: o.ExternalReference, calls: o.Expression[][]) {
+  return o.importExpr(reference, null, span).callFnChain(calls, span);
```
Out of curiosity: do we actually need new output AST logic to handle a function call chain? Can't we just do something like:

```ts
o.importExpr(args).callFn(args1).callFn(args2)....`
```

crisbeto replied:
Ah, you're totally right. I was looking for something like this, but I never thought about chaining the `callFn` calls. That cleaned it up a lot 👍.

kara commented on packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts:
```diff
@@ -180,6 +180,159 @@ describe('compiler compliance: bindings', () => {
+        `;
+
+      expectEmit(result.source, template, 'Incorrect template');
+    });
```
Can we add one more test here that shows multiple nodes in a row with property chaining (perhaps nested)? We probably need a sanity check that property bindings on a child node don't end up in the chain for the parent node.

kara commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1024,21 +1031,23 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-                  [o.literal(input.name),
-                   this.convertPropertyBinding(value, /* skipBindFn */ true)]);
+          propertyBindings.push(
+              {name: input.name, input, value: () => this.convertPropertyBinding(value, true)});
```
Nit: Looks like we lost the inline comment here. Re-add?

```suggestion
              {name: input.name, input, value: () => this.convertPropertyBinding(value, /* skipBindFn */ true)});
```

kara reviewed (approved):
LGTM, with some nits / open questions. Needs an Ivy presubmit since it will affect so many apps.

kara commented on packages/compiler-cli/test/compliance/r3_view_compiler_binding_spec.ts:
```diff
@@ -180,6 +180,159 @@ describe('compiler compliance: bindings', () => {
+            if (rf & 2) {
+              $r3$.ɵɵattribute("id", 2);
+              $r3$.ɵɵpropertyInterpolate("aria-label", 1 + 3);
+              $r3$.ɵɵproperty("title", 1)("tabindex", 3);
```
It seems like we have two options here: 

1) Always optimize for size, and accept template bindings are out of order
2) Always maintain template binding order, and accept size de-optimization in certain cases

I am leaning toward #1 since binding order is technically an implementation detail (and one we are changing from VE to Ivy anyway), but we should think about whether this complicates the mental.

I'll run an Ivy presubmit first to see how break-y it might be...

kara replied:
Ivy presubmit looks normal

kara commented:
[VE presubmit](https://test.corp.google.com/ui#id=OCL:253888438:BASE:253888477:1560899323023:44a8f004)

[Ivy presubmit](https://test.corp.google.com/ui#id=OCL:253888438:BASE:253888835:1560899454158:9cb46c97)

crisbeto committed:
perf(ivy): chain multiple property instructions

Currently each property binding generates an instruction like this:

```
property('prop1', ctx.value1);
property('prop2', ctx.value2);
```

The problem is that we're repeating the call to `property` for each of the properties. Since the `property` instruction returns itself, we can chain all of the calls which is more compact and it looks like this:

```
property('prop1', ctx.value1)('prop2', ctx.value2);
```

These changes implement the chaining behavior for regular property bindings and for synthetic ones, however interpolated ones are still handled like before, because they use a separate instruction.

This PR resolves FW-1389.

crisbeto commented:
Reworked based on the feedback.

kara commented:
merge-assistance: global approval
